rocksdb_admin/admin_handler.cpp

```diff
@@ -929,6 +974,177 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
+  return getHostName() + '_' + topic_name + '_' + partition_id;
+}
+
+void AdminHandler::async_tm_startMessageIngestion(
+    std::unique_ptr<apache::thrift::HandlerCallback<std::unique_ptr<
+      StartMessageIngestionResponse>>> callback,
+    std::unique_ptr<StartMessageIngestionRequest> request) {
+
+  admin::AdminException e;
+  e.errorCode = AdminErrorCode::DB_ADMIN_ERROR;
+
+  const auto db_name = request->db_name;
+  const auto topic_name = request->topic_name;
+  const auto kafka_broker_serverset_path = request->kafka_broker_serverset_path;
+  const auto replay_timestamp_ms = request->replay_timestamp_ms;
+
+  LOG(INFO) << "Called startMessageIngestion for " << db_name;
+
+  // Check if there's already a thread consuming the same partition.
+  if (sync_var_map.find(db_name) != sync_var_map.end()) {
+    e.message = db_name + " is already being consumed";
+    callback.release()->exceptionInThread(std::move(e));
+    LOG(ERROR) << "Already consuming messages to " << db_name <<
+               "in another thread";
+    return;
+  }
+
+  // TODO: Compare the latest value in local_meta_db with replay_timestamp_ms
+  // and choose the latest.
+
+  const auto partition_id = getPartitionIDFromDBName(db_name);
+
+  if (partition_id == -1) {
+    e.message = "Invalid db_name: " + db_name;
+    callback.release()->exceptionInThread(std::move(e));
+    LOG(ERROR) << "Could not find partition in db_name" << db_name;
+    return;
+  }
+
+  const std::unordered_set<uint32_t> partition_ids_set({partition_id});
+
+  // Find the watcher for the given path. If none exists, create a new watcher
+  std::shared_ptr<::kafka::KafkaBrokerFileWatcher> kafka_broker_file_watcher;
+
+  const auto map_iter = serverset_path_to_file_watcher_map.find(
+      kafka_broker_serverset_path);
+
+  if (map_iter == serverset_path_to_file_watcher_map.end()) {
+    kafka_broker_file_watcher =
+        std::make_shared<::kafka::KafkaBrokerFileWatcher>(
+            kafka_broker_serverset_path);
+    serverset_path_to_file_watcher_map[kafka_broker_serverset_path] =
+        kafka_broker_file_watcher;
+  } else {
+    kafka_broker_file_watcher = map_iter->second;
+  }
+
+  // Create new synchronization variables for this partition
+  auto sync_variable = std::make_shared<SyncVariables>();
+  sync_var_map[db_name] = sync_variable;
+
+  SCOPE_EXIT { sync_var_map.erase(db_name); };
+
+  const auto kafka_consumer_pool = std::make_shared<::kafka::KafkaConsumerPool>(
+      kKafkaConsumerPoolSize,
+      partition_ids_set,
+      kafka_broker_file_watcher->GetKafkaBrokerList(),
+      std::unordered_set<std::string>({topic_name}),
+      getConsumerGroupId(topic_name, std::to_string(partition_id)),
+      kKafkaConsumerType);
+
+  const auto kafka_watcher = std::make_unique<KafkaWatcher>(
+      kKafkaWatcherName,
+      kafka_consumer_pool,
+      -1); // kafka_init_blocking_consumer_timeout_ms ;
+
+  // With kafka_init_blocking_consumer_timeout_ms set to -1, messages from
```

Comment:
The mutex need to be locked before we start consuming so that the stopper won't touch the SyncVariable before we start waiting

Reply:
Check this document: https://en.cppreference.com/w/cpp/thread/condition_variable

Reply:
Thanks, I'd missed the part where even if we use an atomic variable, a mutex is required.